tc/core/polyhedral/cuda/mapped_scop.cc

```diff
@@ -405,8 +405,292 @@ bool hasOuterSequentialMember(
+  } else if (nChildren > 1) {
+    for (size_t i = 0; i < nChildren; ++i) {
+      us = us.unite(intersectMappingFilterParams(children[i], us));
+    }
+  }
+
+  return us;
+}
+
+// Change the name of the isl ids tied to threads and blocks
+// by adding a suffix
+isl::union_set modifyMappingNames(
+    isl::union_set set,
+    const std::string suffix) {
+  USING_MAPPING_SHORT_NAMES(BX, BY, BZ, TX, TY, TZ);
+  std::unordered_set<isl::id, isl::IslIdIslHash> identifiers {
+      BX,BY,BZ,TX,TY,TZ};
+
+  auto space = set.get_space();
+  for(auto id : identifiers) {
+    auto name = id.to_str();
+    auto dim = space.find_dim_by_id(isl::dim_type::param, id);
+    CHECK_LE(0, dim);
+    space = space.set_dim_name(isl::dim_type::param, dim, name + suffix);
+  }
+  auto newSet = isl::union_set::empty(space);
+  set.foreach_set([&newSet, &identifiers, &suffix](isl::set set) {
+    for(auto id : identifiers) {
+      auto name = id.to_str();
+      auto dim = set.find_dim_by_id(isl::dim_type::param, id);
+      CHECK_LE(0, dim);
+      set = set.set_dim_name(isl::dim_type::param, dim, name + suffix);
+    }
+    newSet = newSet.unite(set);
+  });
+  return newSet;
+}
+
+// Get the formula computing the index of a thread in a block.
+isl::aff getThreadIdxFormula(
+    isl::space space,
+    const Block& block,
+    const std::string& suffix = "") {
+  USING_MAPPING_SHORT_NAMES(BX, BY, BZ, TX, TY, TZ);
+  std::vector<std::pair<isl::id, unsigned>> mappingIds {
+      {TX, TX.mappingSize(block)},
+      {TY, TY.mappingSize(block)},
+      {TZ, TZ.mappingSize(block)}};
+
+  isl::val zero = isl::val(space.get_ctx(), 0);
+  isl::aff formula = isl::aff(isl::local_space(space), zero);
+
+  for(int i = (int)mappingIds.size() - 1; i >= 0; --i) {
+    auto name = mappingIds[i].first.to_str();
+    auto dim = space.find_dim_by_name(isl::dim_type::param, name + suffix);
+    CHECK_LE(0, dim);
+    auto id = space.get_dim_id(isl::dim_type::param, dim);
+    isl::aff aff(isl::aff::param_on_domain_space(space, id));
+    formula = formula * mappingIds[i].second + aff;
+  }
+
+  return formula;
+}
+
+// Return the constraints ensuring that the points with parameters
+// [t0,t1,t2] and [t0',t1',t2'] are in the same warp.
+// (where t0 is "t0" + suffix1 and t0' is "t0" + suffix2)
+isl::set getSameWarpConstraints(
+    isl::space space,
+    const std::string& suffix1,
+    const std::string& suffix2,
+    const unsigned warpSize,
+    const Block& block) {
+  USING_MAPPING_SHORT_NAMES(BX, BY, BZ, TX, TY, TZ);
+  std::vector<std::pair<isl::id, unsigned>> mappingIds {
+    {TX, TX.mappingSize(block)},
+    {TY, TY.mappingSize(block)},
+    {TZ, TZ.mappingSize(block)}};
+
+  auto formula1 = getThreadIdxFormula(space, block, suffix1);
+  auto formula2 = getThreadIdxFormula(space, block, suffix2);
+
+  return (isl::aff_set((formula1 / warpSize).floor()) == (formula2 / warpSize).floor());
+  }
 } // namespace
 
+int MappedScop::findBestSync(
```

Comment:
This function assumes it is called above thread mapping (and below block mapping).  I would document this assumption.